mongo/integration/unified/testrunner_operation.go

```diff
@@ -111,11 +137,90 @@ func executeTestRunnerOperation(ctx context.Context, operation *operation) error
 		coll := lookupString(args, "collectionName")
 		index := lookupString(args, "indexName")
 		return verifyIndexExists(ctx, db, coll, index, false)
+	case "loop":
+		var unmarshaledArgs loopArgs
+		if err := bson.Unmarshal(args, &unmarshaledArgs); err != nil {
+			return fmt.Errorf("error unmarshalling arguments to loopArgs: %v", err)
+		}
+		return executeLoop(ctx, &unmarshaledArgs, loopDone)
 	default:
 		return fmt.Errorf("unrecognized testRunner operation %q", operation.Name)
 	}
 }
 
+func executeLoop(ctx context.Context, args *loopArgs, loopDone <-chan struct{}) error {
+	// setup entities
+	entityMap := entities(ctx)
+	if args.errorsStored() {
+		if err := entityMap.addBSONArrayEntity(args.ErrorsEntityID); err != nil {
+			return err
+		}
+	}
+	if args.failuresStored() {
+		if err := entityMap.addBSONArrayEntity(args.FailuresEntityID); err != nil {
+			return err
+		}
+	}
+	if args.successesStored() {
+		if err := entityMap.addSuccessesEntity(args.SuccessesEntityID); err != nil {
+			return err
+		}
+	}
+	if args.iterationsStored() {
+		if err := entityMap.addIterationsEntity(args.IterationsEntityID); err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-loopDone:
+			return nil
+		default:
+			if args.iterationsStored() {
+				entityMap.incrementIterations(args.IterationsEntityID)
+			}
+			var loopErr error
+			for i, operation := range args.Operations {
+				if operation.Name == "loop" {
+					return fmt.Errorf("loop sub-operations should not include loop")
+				}
+				loopErr = operation.execute(ctx, loopDone)
+
+				// if the operation errors, stop this loop
+				if loopErr != nil {
+					// If StoreFailures or StoreErrors is set, continue looping, otherwise break
+					if !args.errorsStored() && !args.failuresStored() {
+						return fmt.Errorf("error running loop operation %v : %v", i, loopErr)
+					}
+					errDoc := bson.Raw(bsoncore.NewDocumentBuilder().
+						AppendString("error", loopErr.Error()).
+						AppendInt64("time", getSecondsSinceEpoch()).
+						Build())
+					switch {
+					case !args.errorsStored(): // store errors as failures if storeErrorsAsEntity isnt specified
+						entityMap.appendBSONArrayEntity(args.FailuresEntityID, errDoc)
+					case !args.failuresStored(): // store failures as errors if storeFailuressAsEntity isnt specified
+						entityMap.appendBSONArrayEntity(args.ErrorsEntityID, errDoc)
+					// errors are test runner errors
+					// TODO GODRIVER-1950: use error types to determine error vs failure instead of depending on the fact that
+					// operation.execute prepends "execution failed" to test runner errors
+					case strings.Contains(loopErr.Error(), "execution failed: "):
+						entityMap.appendBSONArrayEntity(args.ErrorsEntityID, errDoc)
+					// failures are if an operation returns an incorrect result or error
+					default:
+						entityMap.appendBSONArrayEntity(args.FailuresEntityID, errDoc)
+					}
+					break
```

Comment:
Why do we `break` here? I assumed we'd want to move on to the next loop operation with `continue`.